### PR TITLE
sink (ticdc): Fix goroutine leak in mysql sink.

### LINKED ISSUE #5356: Goroutine leak in mysql sink
### Which jobs are flaking?

leak_test

### Which test(s) are flaking?

mysql_test

### Jenkins logs or GitHub Actions link

```log
[2022-05-07T06:20:02.424Z] PASS

[2022-05-07T06:20:02.424Z] goleak: Errors on successful test run: found unexpected goroutines:

[2022-05-07T06:20:02.424Z] [Goroutine 349 in state semacquire, with sync.runtime_Semacquire on top of the stack:

[2022-05-07T06:20:02.424Z] goroutine 349 [semacquire]:

[2022-05-07T06:20:02.424Z] sync.runtime_Semacquire(0xc000c8e3b0?)

[2022-05-07T06:20:02.424Z] 	/usr/local/go/src/runtime/sema.go:56 +0x25

[2022-05-07T06:20:02.424Z] sync.(*WaitGroup).Wait(0xc000c8e3a8)

[2022-05-07T06:20:02.424Z] 	/usr/local/go/src/sync/waitgroup.go:136 +0x9c

[2022-05-07T06:20:02.425Z] github.com/pingcap/tiflow/cdc/sink/mysql.(*mysqlSink).notifyAndWaitExec.func1()

[2022-05-07T06:20:02.425Z] 	/home/jenkins/agent/workspace/cdc_ghpr_leak_test/go/src/github.com/pingcap/tiflow/cdc/sink/mysql/mysql.go:456 +0x45

[2022-05-07T06:20:02.425Z] created by github.com/pingcap/tiflow/cdc/sink/mysql.(*mysqlSink).notifyAndWaitExec

[2022-05-07T06:20:02.425Z] 	/home/jenkins/agent/workspace/cdc_ghpr_leak_test/go/src/github.com/pingcap/tiflow/cdc/sink/mysql/mysql.go:455 +0x31b

[2022-05-07T06:20:02.425Z] ]

[2022-05-07T06:20:02.425Z] FAIL	github.com/pingcap/tiflow/cdc/sink/mysql	8.547s
```

https://ci2.pingcap.net/blue/organizations/jenkins/cdc_ghpr_leak_test/detail/cdc_ghpr_leak_test/25186/pipeline

### Anything else we need to know

- Does this test exist for other branches as well?

- Has there been a high frequency of failure lately?

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #5356 

### What is changed and how it works?
**Invesgation:**
1. When `mysqlSink.Close()` is called, all `mysqlWorker` will exit due to the `ctx` is canceled.  At the same time if [this line](https://github.com/pingcap/tiflow/pull/5400/files#diff-ba27cb35a4f83fcd923c1403ff1fb7c1ad29659e7eadd87dd505700bf82f92b9R462) is excuted, the `flushSyncWg.Add(1)` was called but a corespondent `flushSyncWg.Done()` will never be called since all mysqlWorker was exited. So [this goroutine](https://github.com/pingcap/tiflow/pull/5400/files#diff-ba27cb35a4f83fcd923c1403ff1fb7c1ad29659e7eadd87dd505700bf82f92b9R468) will wait for a never done `flushSyncWg`.

**Fix:**
1. To fix this problem, we should make [this goroutine](https://github.com/pingcap/tiflow/pull/5400/files#diff-ba27cb35a4f83fcd923c1403ff1fb7c1ad29659e7eadd87dd505700bf82f92b9R468) be aware of `ctx`'s cancellation, so that it can return at the very time.
2. Add a `workerWg` field in `mysqlSink` to manage all goroutines spawn from `mysqlSink` , and wait all goroutines it spwaned to exit when `mysqlSink.close()` is called.

### Check List 

Tests 

 - Unit test
 - Manual test (add detailed scripts or steps below)

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch

### Release note 

```release-note
Fix a bug that may cause goroutine leak in mysqlSink when sink is closed.
```

### CODE REVIEW

File: cdc/sink/mysql/mysql.go [hunk 1]
```diff
@@ -214,7 +218,12 @@ func NewMySQLSink(
 	if err != nil {
 		return nil, err
 	}
-	go sink.flushRowChangedEvents(ctx, receiver)
+
+	sink.workerWg.Add(1)
+	go func() {
+		sink.flushRowChangedEvents(ctx, receiver)
+		sink.workerWg.Done()
```
Comment (amyangfei) [inline cdc/sink/mysql/mysql.go:225]: It is better to call wg.Done() with a defer, since the defer function will always be called, even `panic` or `runtime.Goexit` happens in `sink.flushRowChangedEvents`
```go
defer sink.workerWg.Done()
sink.flushRowChangedEvents(ctx, receiver)
```

File: cdc/sink/mysql/mysql.go [hunk 2]
```diff
@@ -434,6 +444,7 @@ func (s *mysqlSink) createSinkWorkers(ctx context.Context) error {
 				}
 			}
 			worker.cleanup()
+			s.workerWg.Done()
```
Comment (amyangfei) [inline cdc/sink/mysql/mysql.go:447]: ditto

File: cdc/sink/mysql/mysql_worker.go [hunk 3]
```diff
@@ -81,20 +81,6 @@ func (w *mysqlSinkWorker) run(ctx context.Context) (err error) {
 		txnNum     int
 	)
 
-	// mark FinishWg before worker exits, all data txns can be omitted.
```
Comment (amyangfei) [inline cdc/sink/mysql/mysql_worker.go:84]: why is this removed?

Comment (CharlesCheung96) [reply cdc/sink/mysql/mysql_worker.go:84]: Should we call cleanup here instead of calling it externally？

File: cdc/sink/mysql/mysql.go (hunk 1 above)
Comment (asddongmen) [reply cdc/sink/mysql/mysql.go:225]: Thank for your advice~

File: cdc/sink/mysql/mysql_worker.go (hunk 3 above)
Comment (asddongmen) [reply cdc/sink/mysql/mysql_worker.go:84]: `worker.cleanup` do the same job here:https://github.com/pingcap/tiflow/pull/5400/files#diff-ba27cb35a4f83fcd923c1403ff1fb7c1ad29659e7eadd87dd505700bf82f92b9R446 

So this defer func here seems redundant.

Comment (asddongmen) [reply cdc/sink/mysql/mysql_worker.go:84]: I have moved the `worker.cleanup` into `worker.run`, please help me review again. @amyangfei @CharlesCheung96

File: cdc/sink/mysql/mysql.go [hunk 4]
```diff
@@ -452,10 +462,23 @@ func (s *mysqlSink) notifyAndWaitExec(ctx context.Context) {
...
 	s.execWaitNotifier.Notify()
 	done := make(chan struct{})
+
+	// spawn goroutine from s.workerWg to avoid goroutine leak when mysqlSink is closed.
+	s.workerWg.Add(1)
 	go func() {
-		s.flushSyncWg.Wait()
-		close(done)
+		defer s.workerWg.Done()
+		for {
```
Comment (CharlesCheung96) [inline cdc/sink/mysql/mysql.go:470]: Why do we need a for loop here?

Comment (CharlesCheung96) [inline cdc/sink/mysql/mysql.go:470]: Why do we need a for loop here? Maybe you don't need to call `ctx.Done`.

File: cdc/sink/mysql/mysql.go [hunk 5]
```diff
@@ -452,10 +462,23 @@ func (s *mysqlSink) notifyAndWaitExec(ctx context.Context) {
...
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				s.flushSyncWg.Wait()
```
Comment (CharlesCheung96) [inline cdc/sink/mysql/mysql.go:475]: In the scenario we analyzed earlier, this goroutine will still be stuck here.

> When mysqlSink.Close() is called, all mysqlWorker will exit due to the ctx is canceled. At the same time if [this line](https://github.com/pingcap/tiflow/pull/5400/files#diff-ba27cb35a4f83fcd923c1403ff1fb7c1ad29659e7eadd87dd505700bf82f92b9R462) is excuted, the flushSyncWg.Add(1) was called but a corespondent flushSyncWg.Done() will never be called since all mysqlWorker was exited. So [this goroutine](https://github.com/pingcap/tiflow/pull/5400/files#diff-ba27cb35a4f83fcd923c1403ff1fb7c1ad29659e7eadd87dd505700bf82f92b9R468) will wait for a never done flushSyncWg.

**Comment by hi-rustin** [comment]
/uncc

Recc me when this PR is ready.
